File: go/0169_majority_element.go
func majorityElement(nums []int) int {
	res, cnt := nums[0], 1
	for i := 1; i < len(nums); i++ {
		if nums[i] != res {
			cnt--
			if cnt == 0 {
				res, cnt = nums[i], 1
			}
			continue
		}
		cnt++
	}
	return res
}

File: python/0169_majority_element.py
class Solution:
    def majorityElement(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        candidate = None
        count = 0
        for num in nums:
            if candidate is None:
                candidate = num
                count = 1
                continue

            if num == candidate:
                count += 1
            else:
                count -= 1
                if count == 0:
                    candidate = None
        return candidate